DataFormats/PatCandidates/interface/PATObject.h

```diff
@@ -717,123 +726,167 @@ namespace pat {
-        return overlapItems_[match - overlapLabels_.begin()];
+        
+        auto match = std::lower_bound(overlapLabels_.cbegin(), overlapLabels_.cend(), label);
+        if( match == overlapLabels_.cend() || *match != label ) return pat_statics::EMPTY_CPV;
+        return overlapItems_[std::distance(overlapLabels_.begin(),match)];
   }
 
   template <class ObjectType>
   void PATObject<ObjectType>::setOverlaps(const std::string &label, const reco::CandidatePtrVector & overlaps) {
-        if (!overlaps.empty()) {
-            std::vector<std::string>::const_iterator match = std::find(overlapLabels_.begin(), overlapLabels_.end(), label);
-            if (match == overlapLabels_.end()) {
-                overlapLabels_.push_back(label);
-                overlapItems_.push_back(overlaps);
-            } else {
-                overlapItems_[match - overlapLabels_.begin()] = overlaps;
-            }
-        }
+      
+        auto match = std::lower_bound(overlapLabels_.begin(), overlapLabels_.end(), label);
+        const auto dist = std::distance(overlapLabels_.begin(),match);
+        if ( match == overlapLabels_.end() || *match != label ) {
+          overlapLabels_.insert(match,label);
+          overlapItems_.insert(overlapItems_.begin()+dist,overlaps);
+        } else {
+          overlapItems_[dist] = overlaps;
+        }        
   }
 
   template <class ObjectType>
   const pat::UserData * PATObject<ObjectType>::userDataObject_( const std::string & key ) const
   {
-    std::vector<std::string>::const_iterator it = std::find(userDataLabels_.begin(), userDataLabels_.end(), key);
-    if (it != userDataLabels_.end()) {
-        return & userDataObjects_[it - userDataLabels_.begin()];
+    auto it = std::lower_bound(userDataLabels_.cbegin(),userDataLabels_.cend(),key);
+    if ( it != userDataLabels_.cend() && *it == key) {
+      return &userDataObjects_[std::distance(userDataLabels_.cbegin(),it)];
     }
-    return 0;
+    return nullptr;
   }
 
   template <class ObjectType>
   float PATObject<ObjectType>::userFloat( const std::string &key ) const
   {
-    std::vector<std::string>::const_iterator it = std::find(userFloatLabels_.begin(), userFloatLabels_.end(), key);
-    if (it != userFloatLabels_.end()) {
-        return userFloats_[it - userFloatLabels_.begin()];
+    auto it = std::lower_bound(userFloatLabels_.cbegin(),userFloatLabels_.cend(),key);
+    if( it != userFloatLabels_.cend() && *it == key ) {
+      return userFloats_[std::distance(userFloatLabels_.cbegin(),it)];
     }
-    return 0.0;
+    // cuts based on non-existing keys should always fail, so use NaN, not zero
+    cms::Exception ex("UnknownUserFloat");
+    ex << "Requested UserFloat " << key << " is not available! Possible UserFloats are: " << std::endl;
+    for( const auto& name : userFloatLabels_ ) {
+      ex << name << ' ';
+    }
+    throw ex;
+    return std::numeric_limits<float>::quiet_NaN(); 
```

Comment:
it may be better to add a non-templated method `pat::throwMissingLabel(const std::string & what, const std::string &value, const std::vector<std::string> &availableValues)` that does the building of the exception string and and throwing of it, so that the code here can just be `throwMissingLabel("UserFloat", key, userFloatLabels_)`.